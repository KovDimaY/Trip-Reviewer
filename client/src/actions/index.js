import axios from 'axios';

import {
    GET_TRIPS,
    GET_TRIP_W_REVIEWER,
    CLEAR_TRIP_W_REVIEWER,
    USER_LOGIN,
    USER_AUTH,
    ADD_TRIP,
    CLEAR_NEW_TRIP,
    GET_USER_REVIEWS
} from '../constants/action-names';

export function getTrips(
    limit = 10,
    start = 0,
    order = 'asc',
    list = []
) {
    const request = axios
        .get(`/api/getManyTrips?limit=${limit}&skip=${start}&order=${order}`)
        .then(response => [...list, ...response.data]);

    return {
        type: GET_TRIPS,
        payload: request
    };
}

export function getTripWithReviewer(id) {
    const request = axios.get(`/api/getTrip?id=${id}`);

    return (dispatch) => {
        request.then(({ data }) => {
            const trip = data;

            axios
                .get(`/api/getReviewer?id=${trip.ownerId}`)
                .then(({ data }) => {
                    const response = {
                        trip,
                        reviewer: data
                    };

                    dispatch({
                        type: GET_TRIP_W_REVIEWER,
                        payload: response
                    });
                });
        });
    };
}

export function getUserReviews(userId) {
    const request = axios
        .get(`/api/getUserReviews?user=${userId}`)
        .then(response => response.data);

    return {
        type: GET_USER_REVIEWS,
        payload: request
    };
}

export function clearTripWithReviewer() {
    return {
        type: CLEAR_TRIP_W_REVIEWER,
        payload: {
            trip: null,
            reviewer: null
        }
    };
}

export function loginUser({ email, password }) {
    const request = axios
        .post('/api/login',{ email, password })
        .then(response => response.data);

    return {
        type: USER_LOGIN,
        payload: request
    };
}

export function auth() {
    const request = axios
        .get('/api/auth')
        .then(response => response.data);

    return {
        type: USER_AUTH,
        payload: request
    }
}

export function addTrip(trip) {
    const request = axios
        .post('/api/trip', trip)
        .then(response => response.data);

    return {
        type: ADD_TRIP,
        payload: request
    };
}

export function clearNewTrip() {
    return {
        type: CLEAR_NEW_TRIP,
        payload: null
    };
}