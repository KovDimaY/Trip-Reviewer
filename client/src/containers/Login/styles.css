.rl_container .field-error {
    border: 2px solid red;
    color: red;
}