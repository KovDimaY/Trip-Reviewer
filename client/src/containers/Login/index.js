import React, { Component } from 'react';
import { connect } from 'react-redux';

import { loginUser } from '../../actions';

import './styles.css';

class Login extends Component {
    state = {
        formData: {
            email: '',
            password: ''
        },
        hideError: {
            email: false,
            password: false
        }
    };

    handleInput = (event) => {
        const newFormData = { ...this.state.formData };
        const newHideError = { ...this.state.hideError };
        const { value, name } = event.target;

        newFormData[name] = value;
        newHideError[name] = true;

        this.setState({
            formData: newFormData,
            hideError: newHideError
        });
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.users.login.isAuth) {
            this.props.history.push('/user');
        }
        this.setState({
            hideError: {
                email: false,
                password: false
            }
        });
    }

    submitForm = (event) => {
        event.preventDefault();
        this.props.dispatch(loginUser(this.state.formData));
    }

    getErrorClass(param) {
        const { error } = this.props.users.login;
        const { hideError } = this.state;

        if (error && error.field === param && !hideError[param]) {
            return 'field-error';
        }
        return '';
    }

    renderError(param) {
        const { error } = this.props.users.login;
        const { hideError } = this.state;

        if (error && error.field === param && !hideError[param]) {
            return <div className="error">{error.message}</div>;
        }
    }

    render() {
        const { error } = this.props.users.login;
        const { email, password } = this.state.formData;

        return (
            <div className="login-container">
                <form onSubmit={this.submitForm}>
                    <h2>Log in here</h2>

                    <div className="form_element">
                        <input 
                            type="email"
                            name="email"
                            className={this.getErrorClass('email')}
                            placeholder="Enter your mail"
                            value={email}
                            onChange={this.handleInput}
                        />
                    </div>
                    { this.renderError('email') }

                    <div className="form_element">
                        <input 
                            type="password"
                            name="password"
                            className={this.getErrorClass('password')}
                            placeholder="Enter your password"
                            value={password}
                            onChange={this.handleInput}
                        />
                    </div>
                    { this.renderError('password') }

                    <button type="submit">Log in</button><br/>

                    {
                        error && error.message && <a href="reset-password" className="reset-password">Forgot my password</a>
                    }

                </form>
            </div>
        );
    }
}

function mapStateToProps(state) {
    return {
        users: state.users
    };
}

export default connect(mapStateToProps)(Login);