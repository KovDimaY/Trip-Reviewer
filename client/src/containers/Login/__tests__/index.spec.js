import React from 'react';
import configureStore from 'redux-mock-store';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';
 
import Login from './../../Login';
import { loginUser } from './../../../actions';

const mockStore = configureStore();

jest.mock('./../../../actions', () => ({ 
    loginUser: jest.fn(() => ({
        type: 'loginUser'
    })) 
}));

const mockComponent = (initialState = {}, props) => {
    const store = mockStore(initialState);

    return (
        <Login {...props} store={store} />
    );
};
 
describe('<Login />', () => {
    it('should render component', () => {
        const initialState = {
            users: {}
        };
        const tree = create(mockComponent(initialState)).toJSON();
        
        expect(tree).toMatchSnapshot();
    });

    it('should render component with error message', () => {
        const initialState = {
            users: {
                login: {
                    message: 'test'
                }
            }
        };
        const tree = create(mockComponent(initialState)).toJSON();
        
        expect(tree).toMatchSnapshot();
    });

    it('should change state when handleInputEmail is called', () => {
        const initialState = {
            users: {}
        };
        const event = {
            target: {
                value: 'test'
            }
        };

        const instance = shallow(mockComponent(initialState)).dive().instance();
    
        instance.handleInputEmail(event);
    
        expect(instance.state.email).toEqual('test');
    });

    it('should change state when handleInputPassword is called', () => {
        const initialState = {
            users: {}
        };
        const event = {
            target: {
                value: 'test'
            }
        };

        const instance = shallow(mockComponent(initialState)).dive().instance();
    
        instance.handleInputPassword(event);
    
        expect(instance.state.password).toEqual('test');
    });

    it('should dispatch loginUser when submitForm is called', () => {
        const preventDefault = jest.fn();
        const state = {
            email: 'test',
            password: 'test'
        };
        const initialState = {
            users: {}
        };
        const event = { preventDefault };

        const instance = shallow(mockComponent(initialState)).dive().instance();
        instance.setState(state);
    
        instance.submitForm(event);

        expect(preventDefault).toHaveBeenCalled();    
        expect(loginUser).toHaveBeenCalledWith(state);
    });

    it('should redirect when componentWillReceiveProps is called and authenticated', () => {
        const push = jest.fn();
        const props = {
            history: {
                push
            }
        };
        const nextProps = {
            users: {
                login: {
                    isAuth: true
                }
            }
        };
        const initialState = {
            users: {}
        };

        const instance = shallow(mockComponent(initialState, props)).dive().instance();
    
        instance.componentWillReceiveProps(nextProps);

        expect(push).toHaveBeenCalledWith('/user');
    });

    it('should not redirect when componentWillReceiveProps is called and not authenticated', () => {
        const push = jest.fn();
        const props = {
            history: {
                push
            }
        };
        const nextProps = {
            users: {
                login: {}
            }
        };
        const initialState = {
            users: {}
        };

        const instance = shallow(mockComponent(initialState, props)).dive().instance();
    
        instance.componentWillReceiveProps(nextProps);

        expect(push).not.toHaveBeenCalledWith('/user');
    });
});