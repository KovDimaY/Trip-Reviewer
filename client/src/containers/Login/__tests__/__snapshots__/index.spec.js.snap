// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should render component 1`] = `
<div
  className="rl_container"
>
  <form
    onSubmit={[Function]}
  >
    <h2>
      Log in here
    </h2>
    <div
      className="form_element"
    >
      <input
        onChange={[Function]}
        placeholder="Enter your mail"
        type="email"
        value=""
      />
    </div>
    <div
      className="form_element"
    >
      <input
        onChange={[Function]}
        placeholder="Enter your password"
        type="password"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      Log in
    </button>
    <div
      className="error"
    />
  </form>
</div>
`;

exports[`<Login /> should render component with error message 1`] = `
<div
  className="rl_container"
>
  <form
    onSubmit={[Function]}
  >
    <h2>
      Log in here
    </h2>
    <div
      className="form_element"
    >
      <input
        onChange={[Function]}
        placeholder="Enter your mail"
        type="email"
        value=""
      />
    </div>
    <div
      className="form_element"
    >
      <input
        onChange={[Function]}
        placeholder="Enter your password"
        type="password"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      Log in
    </button>
    <div
      className="error"
    >
      <div>
        test
      </div>
    </div>
  </form>
</div>
`;
